src/otf_collector.cpp

```diff
@@ -0,0 +1,232 @@
+    // Spawn IBF membership agent in this scope because it is expensive
+    auto && ibf_ref = ibf.getIBF();
+    auto agent = ibf_ref.membership_agent();
+
+    CollectionItem queue_init(NFA, 0, hit_vector, 0, 0);
+    queue.emplace(queue_init);
+
+    State *current_state;
+    kmer_t kmer;
+    path_t current_path;
+    uint8_t split_hits;
+    uint8_t shift_counts;
+
+    CollectionItem item1;
+    CollectionItem item2;
+
+    while(!queue.empty())
+    {
+        current_state = queue.front().nfa_state_;
+        kmer = queue.front().kmer_;
+        current_path = queue.front().path_;
+        split_hits = queue.front().cycles_;
+        shift_counts = queue.front().shift_count_;
+        queue.pop();
+        
+        int symbol = current_state->c_;
+        switch(symbol)
+        {
+            case Match:
+                path_matrix.raw_data() |= current_path.raw_data();
+                break;
+            case SplitU:
+                item1 = {current_state->out1_, kmer, current_path, split_hits, shift_counts};
+                item2 = {current_state->out2_, kmer, current_path, split_hits, shift_counts};
+                if((current_state->out1_->c_ > SplitU)) item1.cycles_++;
+                if((current_state->out2_->c_ > SplitU)) item2.cycles_++;
+                queue.push(item1);
+                queue.push(item2);
+                break;
+            case SplitP:
+                if(split_hits == (ibf.k_-1)) // The k-1 removes redundant homopolymer kmers
+                {
+                    item1 = {current_state->out1_, kmer, current_path, 0, shift_counts};
+                    queue.push(item1);
+                    break;
+                }
+                item1 = {current_state->out1_, kmer, current_path, 0, shift_counts};
+                item2 = {current_state->out2_, kmer, current_path, split_hits, shift_counts};
+                queue.push(item1);
+                queue.push(item2);
+                break;
+            case SplitK:
+                if(split_hits == ibf.k_)
+                {
+                    item1 = {current_state->out1_, kmer, current_path, 0, shift_counts};
+                    queue.push(item1);
+                    break;
+                }
+                item1 = {current_state->out1_, kmer, current_path, 0, shift_counts};
+                item2 = {current_state->out2_, kmer, current_path, split_hits, shift_counts};
+                queue.push(item1);
+                queue.push(item2);
+                break;
+            default:
+                if(current_state->out1_->c_ > SplitU) split_hits++;
+                update_path(kmer, shift_counts, symbol, agent, current_path, ibf, kmer_cache);
+                if(all_bits_zero(current_path)) break; // Immediately get rid of deadend paths
+                item1 = {current_state->out1_, kmer, current_path, split_hits, shift_counts};
+                queue.push(item1);
+                break;
+        }
+    }
```

Comment:
It is nice to declare the variables at the point, where they are needed:
( I havn't tested this code)

```suggestion
    while(!queue.empty())
    {
        auto front = std::move(queue.front()); // Moving the data out of the queue
        queue.pop(); // removing the remaining data from the queue
        
        State* current_state = front.nfa_state_;
        kmer_t const& kmer = front.kmer_;
        path_t const& current_path = front.path_;
        uint8_t split_hits = front.cycles_;
        uint8_t shift_counts = front.shift_count_;
        
        int symbol = current_state->c_;
        switch(symbol)
        {
            case Match:
                path_matrix.raw_data() |= current_path.raw_data();
                break;
            case SplitU:
                queue.emplace(current_state->out1_, kmer, current_path, split_hits, shift_counts);
                if((current_state->out1_->c_ > SplitU)) queue.back().cycles_++;
                queue.emplace(current_state->out2_, kmer, current_path, split_hits, shift_counts);
                if((current_state->out2_->c_ > SplitU)) queue.back().cycles_++;
                break;
            case SplitP:
                if(split_hits == (ibf.k_-1)) // The k-1 removes redundant homopolymer kmers
                {
                    queue.emplace(current_state->out1_, kmer, current_path, 0, shift_counts);
                    break;
                }
                queue.emplace(current_state->out1_, kmer, current_path, 0, shift_counts);
                queue.emplace(current_state->out2_, kmer, current_path, split_hits, shift_counts);
                break;
            case SplitK:
                if(split_hits == ibf.k_)
                {
                    queue.emplace(current_state->out1_, kmer, current_path, 0, shift_counts);
                    break;
                }
                queue.emplace(current_state->out1_, kmer, current_path, 0, shift_counts);
                queue.emplace(current_state->out2_, kmer, current_path, split_hits, shift_counts);
                break;
            default:
                if(current_state->out1_->c_ > SplitU) split_hits++;
                update_path(kmer, shift_counts, symbol, agent, current_path, ibf, kmer_cache);
                if(all_bits_zero(current_path)) break; // Immediately get rid of deadend paths
                queue.emplace(current_state->out1_, kmer, current_path, split_hits, shift_counts);
                break;
        }
    }
```